discord.py 2.0 update (3d914e08->2.0.1)

Jackenmen opened the pull request:
### Description of the changes

The update is long due...

d.py diff: https://github.com/Rapptz/discord.py/compare/3d914e08...v2.0.1

Things to do:

- overall double-checking of changes from d.py diff + cross-check with Danny's weekly changelogs
- consider adding support for command argument descriptions? I think this could probably be done in a separate PR

Open questions:

- Is the proposed approach for handling voice mutes (`[p]voicemute`) and channel mutes in VC (`[p]channelmute`) alright?
  - in this PR, `[p]voicemute` removes only the Speak perm while `[p]channelmute` removes both message sending perms and speak perms

Things that are handled separately:
- hybrid commands business
    - this is done because I think that there's a lot related to it that we have to work on and it doesn't actually make any sense to work on it all in this PR

### Have the changes in this PR been tested?

No

Jackenmen committed:
[TiV] Update type hints to include VoiceChannel where applicable

Jackenmen committed:
[TiV] Allow setting voice channels as output channels

BREAKING:
- List returned by `Red.get_owner_notification_destinations()`
  may now contain VoiceChannel
- `modlog.get_modlog_channel()` may now return VoiceChannel

Rest of the changes:
- Stream alerts can now be sent to voice channels
- These methods now accept VoiceChannel:
  - `modlog.Case.from_json()`
  - `modlog.set_modlog_channel()`
- These commands now accept VoiceChannel:
  - `[p]announceset channel` in admin cog
  - `[p]muteset notification` in mutes cog
  - `[p]reportset output` in reports cog
  - `[p]warningset warnchannel` in warnings cog
  - `[p]modlogset modlog` in core

Jackenmen committed:
[TiV] Allow setting filters for VCs in Filter cog

Jackenmen committed:
[TiV] Prevent usage of `[p]slowmode` in voice channels

Jackenmen committed:
[TiV] Allow passing voice channels to `[p]diagnoseissues`

Jackenmen committed:
[TiV] Allow ignoring (`[p]ignore`) voice channels

Jackenmen committed:
[TiV] Support passing VoiceChannel to `Red.embed_requested()`

Jackenmen committed:
Update Red to work after removal of d.py's get_converter() internal

Jackenmen committed:
Use new dynamic default arguments functionality where possible

Jackenmen committed:
[forums] Update filter channel add/remove to allow passing forums

Jackenmen committed:
[forums] Allow ignoring (`[p]ignore`) forums

Jackenmen committed:
[forums] Allow changing embed setting for forums

Jackenmen committed:
[forums] Update Case.parent_channel type hint

Jackenmen committed:
[forums] Update permissions documentation/type hints for forums

Jackenmen committed:
[TiV] Update permissions documentation to be clear about VCs

Jackenmen committed:
Remove DEP-WARN about discord.utils.MISSING now that it's documented

Jackenmen committed:
Permissions.use_slash_commands->use_application_commands

Jackenmen committed:
Replace isinstance checks on channel with `.guild is (not) None`

This was done because of c22b6dc2cd7cafb3d0e892bd2709992afcc3e941
but we actually have not used anything that it added so
this could have been done in separate PR...

Jackenmen committed:
Use Client.application attribute instead of pre-fetching app ourselves

Jackenmen committed:
Bump discord.py to 2.0.1

Jackenmen committed:
Improve output of test failure for commands re-exports

Zephyrkul commented on redbot/core/commands/__init__.py:
```diff
@@ -189,4 +189,13 @@
+    CurrentChannel as CurrentChannel,
+    Author as Author,
+    param as param,
+    MissingRequiredAttachment,
```
```suggestion
    MissingRequiredAttachment as MissingRequiredAttachment,
```

Jackenmen committed:
Add new perms to commands.requires.DM_PERMS

Jackenmen committed:
Remove dead code in Requires class

Jackenmen committed:
BREAKING: Remove commands.requires.DM_PERMS

Not a single public repository on GitHub seems to be using it
and we don't use it either.

Jackenmen committed:
commands.cooldown's callable now accepts context instead of message

Jackenmen committed:
Use newly added Context.{bot_,}permissions

Jackenmen committed:
Change discord.py's logging level to INFO

A lot of logs at INFO level have been lowered to DEBUG
so this should be saner now?

Jackenmen committed:
Add missing re-exports

Jackenmen committed:
Bump RLL to 0.11.0rc1

Jackenmen committed:
[TiV] Update [p]voicemute to only disable Speak permission

Jackenmen committed:
[TiV] Update MessagePredicate type hints and docs to include VCs

Jackenmen committed:
Temporarily ignore missing app commands stuff in commands module

Work on that is done separately.

Jackenmen committed:
Add use_application_commands perm to denied perms in Mutes cog

TrustyJAID reviewed (dismissed):
I've been testing and using this update for several months on one bot and over 2 weeks on 2 other bots without any issues so far.

Jackenmen commented:
Have you looked through the new commits (all of the ones in Oct 1, 2022 section)? Especially this one:
[`433b3ab` (#5709)](https://github.com/Cog-Creators/Red-DiscordBot/pull/5709/commits/433b3ab019b5cf524c5c02e0c092a2029d77d276)

TrustyJAID commented:
I did look at it and it seems fine. It will require some updates on #5604 since that uses dataclasses instead of a dict but I'll deal with that after this is merged.

Kowlin reviewed (dismissed):
Looks good in my eyes after reading every commit.

leetfin reviewed (commented):
I've been testing with this PR for months now, no issues noted.

Drapersniper commented:
For the folks planning to say they have been testing this for months, the issue at hand is there have been several recent changes - so any testing prior to these (Oct 1, 2022) is not fully relevant

Jackenmen committed:
Minor type hint changes regarding ForumChannels

Jackenmen committed:
Use `Member.get_role(role_id)` instead of `role in Member.roles`

Jackenmen committed:
Update Tunnel.files_from_attach() to preserve attachment descriptions

Kowlin reviewed (approved):
Tested and ran the latest changes, nothing wrong there. You're amazing Jack.
